onnxruntime/hosting/http/session.cc

```diff
@@ -96,23 +94,42 @@ template <typename Body, typename Allocator>
 void HttpSession::HandleRequest(http::request<Body, http::basic_fields<Allocator> >&& req) {
   HttpContext context{};
   context.request = std::move(req);
-
   // TODO: set request id
-  std::string path = context.request.target().to_string();
-  std::string model_name, model_version, action;
-  handler_fn func;
-  http::status status = routes_->ParseUrl(context.request.method(), path, model_name, model_version, action, func);
 
-  if (http::status::ok == status && func != nullptr) {
-    func(model_name, model_version, action, context);
-  } else {
-    context.response.result(status);
+  auto status = ExecuteUserFunction(context);
+
+  if (status != http::status::ok) {
+    routes_->on_error(context);
   }
 
   context.response.keep_alive(context.request.keep_alive());
   context.response.prepare_payload();
   return Send(std::move(context.response));
 }
 
+http::status HttpSession::ExecuteUserFunction(HttpContext& context) {
+  std::string path = context.request.target().to_string();
+  std::string model_name, model_version, action;
+  HandlerFn func;
+
+  auto status = routes_->ParseUrl(context.request.method(), path, model_name, model_version, action, func);
+
+  if (status != http::status::ok) {
+    context.error_code = status;
+    std::string method_str = std::string(http::to_string(context.request.method()));
+    context.error_message = "Path not found with method: " + method_str + " and path: " + context.request.target().to_string();
+    return status;
+  }
+
+  try {
+    func(model_name, model_version, action, context);
+  } catch (const std::exception& ex) {
+    context.error_message = std::string(ex.what());
```

Comment:
Can we make `OnError` support any kind of errors including 500?

Reply:
Yep. Accidentally hard coded the error in the callback when exploring ideas. Fixed.